### PR TITLE
utter end-to-end responses

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- fix https://github.com/RasaHQ/rasa/issues/6410
- deprecate `random_template_for` as it's unused
- rename `Domain.action_names` to `Domain.action_names_or_texts`. Added a property with deprecation message for `Domain.action_names`
- add docstrings
- utter end-to-end bot responses if predicted
- log end-to-end action correctly (`action_text` property is set)
- added method `event_for_successful_execution` to `Action` to encapsulate action event with the action which ran it

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by wochinge** [comment]
@Ghostvv how can I test this? I tried using the `e2e` example bot but it only predicts regular actions 🤯

File: rasa/core/actions/action.py [hunk 1]
```diff
@@ -170,7 +170,7 @@ def action_from_name(
     ):
         return ActionRetrieveResponse(name)
 
-    if name.startswith(UTTER_PREFIX):
+    if name.startswith(UTTER_PREFIX) or name in domain.action_names:
```
Comment (joejuzl) [inline rasa/core/actions/action.py:173]: Won't this prevent remote actions?

File: rasa/shared/core/domain.py [hunk 2]
```diff
@@ -521,12 +536,15 @@ def __init__(
...
 
         self.slots = slots
-        self.templates = templates
-        self.action_texts = action_texts or []
         self.session_config = session_config
 
         self._custom_actions = action_names
 
+        self.action_texts = action_texts or []
+        self._user_responses = templates
```
Comment (joejuzl) [inline rasa/shared/core/domain.py:544]: Does this mean responses _to_ the user?

File: rasa/shared/importers/importer.py [hunk 3]
```diff
@@ -464,7 +464,7 @@ async def _get_domain_with_e2e_actions(self) -> Domain:
             [],
             {},
             action_names=[],
-            forms=[],
+            forms={},
```
Comment (joejuzl) [inline rasa/shared/importers/importer.py:467]: What's the reason for this change?

File: rasa/core/actions/action.py (hunk 1 above)
Comment (wochinge) [reply rasa/core/actions/action.py:173]: omg, yes of course. That was a bling auto complete 🙈

File: rasa/shared/importers/importer.py (hunk 3 above)
Comment (wochinge) [reply rasa/shared/importers/importer.py:467]: That's related to [this change](https://github.com/RasaHQ/rasa/pull/7171). Seems something went wrong when merging `master` into this branch.

File: rasa/shared/core/domain.py (hunk 2 above)
Comment (wochinge) [reply rasa/shared/core/domain.py]: renamed it. Is it clearer now?

Comment (wochinge) [reply rasa/shared/core/domain.py]: The difference is that these responses are defined in the domain by the developer. The bot will send them to the user. The `action_texts` are from stories or rules and will also be sent to the user. However, the developer defines them in an end-to-end fashion within the stories / rules instead of the domain. Does that make sense?

**Comment by Ghostvv** [comment]
>  I tried using the e2e example bot but it only predicts regular actions 

strange, it predicts text for me, if I enter `Hello` or smth

**Comment by wochinge** [comment]
I'll try again later 👍 🤔

**Comment by wochinge** [comment]
@Ghostvv The tests fail due to story printing, correct?

**Comment by Ghostvv** [comment]
yes, this are the tests I was talking about earlier

**Comment by wochinge** [comment]
@joejuzl @Ghostvv Ready for another review. It works locally. Also fixed how the action is logged (we previously logged the action text in the action name property).

Testing with the example bot worked now 🤷

**Comment by joejuzl** [comment]
LGTM

File: rasa/core/actions/action.py [hunk 4]
```diff
@@ -95,17 +97,26 @@ def action_for_index(
...
+        prediction.is_end_to_end_prediction,
+    )
 
 
 def action_for_name(
-    action_name: Text, domain: Domain, action_endpoint: Optional[EndpointConfig]
+    action_name: Text,
+    domain: Domain,
+    action_endpoint: Optional[EndpointConfig],
+    is_end_to_end_prediction: bool = False,
```
Comment (Ghostvv) [inline rasa/core/actions/action.py:119]: in `Prediction` class `is_end_to_end_prediction` means that a prediction was made using user text, is it the same here?
In this case it shouldn't affect whether it is `action_text` or `action_name`, It's domain's job to differentiate based on given index

File: rasa/core/policies/policy.py [hunk 5]
```diff
@@ -460,6 +460,30 @@ def max_confidence(self) -> float:
...
+            action_name: The name of the predicted action. If it was an end-to-end
+                prediction the `action_name` is actually the predicted text.
+
+        Returns:
+            The events which should be logged.
+        """
+        action_text = None
+        if self.is_end_to_end_prediction:
+            action_text = action_name
+            action_name = None
```
Comment (Ghostvv) [inline rasa/core/policies/policy.py:476]: this is incorrect. Whether it is `action_text` or `action_name` decides domain.
`self.is_end_to_end_prediction` only indicates what type of user input (text or intent) was used

**Review by Ghostvv — CHANGES_REQUESTED** [review]

File: rasa/core/policies/policy.py (hunk 5 above)
Comment (wochinge) [reply rasa/core/policies/policy.py]: Yes, of course 😬

Comment (wochinge) [reply rasa/core/policies/policy.py]: Too much refactoring - I could luckily just revert the commit 😁

File: rasa/core/policies/policy.py [hunk 6]
```diff
@@ -16,7 +16,7 @@
     TYPE_CHECKING,
 )
 import numpy as np
-from rasa.shared.core.events import Event
+from rasa.shared.core.events import Event, ActionExecuted
```
Comment (Ghostvv) [inline rasa/core/policies/policy.py:19]: I think `ActionExecuted` will be unused export?

File: rasa/core/processor.py [hunk 7]
```diff
@@ -829,15 +829,26 @@ def _log_action_on_tracker(
...
+            tracker.update(self._action_event_for_prediction(action_name, prediction))
 
         logger.debug(f"Action '{action_name}' ended with events '{events}'.")
         tracker.update_with_events(events, self.domain)
 
+    def _action_event_for_prediction(
+        self, action_name: Text, prediction: PolicyPrediction
+    ) -> ActionExecuted:
+        action_text = None
+        if action_name in self.domain.action_texts:
```
Comment (Ghostvv) [inline rasa/core/processor.py:841]: I would prefer modifying the part of the code (maybe inside the `domain.py`) that transforms predicted index into an action name, to return a tuple `(action_name, action_text)` one of which would be `None`. Then all of these method should take both name and text and pass it along. this way we decouple the logic of text vs name and contain it only within domain file

Comment (wochinge) [reply rasa/core/processor.py:841]: We only pass around the `Action` object which we retrieve [here](https://github.com/RasaHQ/rasa/blob/d43364627ce27b6c7048a7e460a0087a84447f3e/rasa/core/processor.py#L357). I think passing around additional information via tuples rather increases complexity than to help with it. What I could do is to set a flag `is_end_to_end_action` on the `Action` object [here](https://github.com/RasaHQ/rasa/blob/8f60f83d089e0e4dcbd79900c7598935f50119e6/rasa/core/actions/action.py#L173) or even add method `action_executed_for_event` to `Action` which returns the correct `ActionExecuted` event.

Comment (Ghostvv) [reply rasa/core/processor.py:841]: I think `action_for_index` is a perfect place for that. Instead of doing it inside `action_for_name`, create method `action_for_text` and differentiate between them inside `action_for_index`. What do you think?

Comment (wochinge) [reply rasa/core/processor.py:841]: I don't see any advantage of doing this. What are you to accomplish? If you want to isolate determining whether it's an end-to-end action, then I'd go for the described approach above.

Comment (Ghostvv) [reply rasa/core/processor.py:841]: because the name `action_for_name` is confusing. Name doesn't decide the action anymore, it is the index that decides that. Why do we have 2 identical methods on the first place?

Comment (Ghostvv) [reply rasa/core/processor.py:841]: I want to avoid putting `action_text` into variable called `action_name` in a couple of month it'll create terrible confusion

Comment (wochinge) [reply rasa/core/processor.py:841]: > Why do we have 2 identical methods on the first place?

Which two methods are you referring to? Index and name? There are some cases where have a name and want to get the action based on that (e.g. for the `action_session_started)

> I want to avoid putting action_text into variable called action_name in a couple of month it'll create terrible confusion

I see that. As only `_log_action_on_tracker` is having this variable I still suggest doing what I suggested [in my earlier. comment.](https://github.com/RasaHQ/rasa/pull/7325#discussion_r530490382).

Comment (Ghostvv) [reply rasa/core/processor.py:841]: I'm not sure I fully follow your proposal, but if you think it is better let's go for it. I'd prefer the implementation where there is no need for additional flags

Comment (wochinge) [reply rasa/core/processor.py:841]: I agree on the flag thing and I've an idea about that. Can hopefully do a quick draft later which we can then discuss.

File: rasa/core/actions/action.py [hunk 8]
```diff
@@ -147,41 +125,53 @@ def is_retrieval_action(action_name: Text, retrieval_intents: List[Text]) -> boo
...
-    if name.startswith(UTTER_PREFIX) and is_retrieval_action(
-        name, domain.retrieval_intents
+    if action_name_or_text.startswith(UTTER_PREFIX) and is_retrieval_action(
+        action_name_or_text, domain.retrieval_intents
     ):
-        return ActionRetrieveResponse(name)
+        return ActionRetrieveResponse(action_name_or_text)
+
+    if action_name_or_text in domain.action_texts:
+        return EndToEndAction(action_name_or_text)
```
Comment (wochinge) [inline rasa/core/actions/action.py:161]: this is the essential part 1 of the change

File: rasa/core/actions/action.py [hunk 9]
```diff
@@ -244,6 +234,21 @@ async def run(
     def __str__(self) -> Text:
         return "Action('{}')".format(self.name())
 
+    def event_for_successful_execution(
```
Comment (wochinge) [inline rasa/core/actions/action.py:237]: this is the essential part 2 of the change

File: rasa/core/actions/action.py [hunk 10]
```diff
@@ -283,6 +288,51 @@ def __str__(self) -> Text:
         return "ActionUtterTemplate('{}')".format(self.name())
 
 
+class EndToEndAction(Action):
```
Comment (wochinge) [inline rasa/core/actions/action.py:291]: this is the essential part 3 of the change

File: rasa/core/processor.py (hunk 7 above)
Comment (wochinge) [reply rasa/core/processor.py:842]: @Ghostvv Updated it. Way cleaner solution now. I was able to get rid of all domain changes. Haven't polished the PR yet because I wanted to share it with you first, so please comment on the higher level things

File: rasa/core/actions/action.py [hunk 11]
```diff
@@ -147,41 +125,53 @@ def is_retrieval_action(action_name: Text, retrieval_intents: List[Text]) -> boo
...
         domain: The current model domain.
         action_endpoint: The endpoint to execute custom actions.
 
+    Raises:
+        ActionNotFoundException: If action not in current domain.
     Returns:
         The instantiated action.
     """
+
+    if action_name_or_text not in domain.action_names:
```
Comment (Ghostvv) [inline rasa/core/actions/action.py:144]: is it still `domain.action_names`? Can we rename this one as well?

File: rasa/core/actions/action.py (hunk 8 above)
Comment (Ghostvv) [reply rasa/core/actions/action.py:161]: I'd say logically it should go before checking `.startswith(UTTER_PREFIX)` on Line 155

File: rasa/core/actions/action.py (hunk 10 above)
Comment (Ghostvv) [reply rasa/core/actions/action.py:291]: shall we name it in the spirit of other actions? Smth like `ActionEndToEndText`? or smth similar

**Review by Ghostvv — COMMENTED** [review]
I like your approach, I agree it makes things clean

File: rasa/core/actions/action.py (hunk 8 above)
Comment (wochinge) [reply rasa/core/actions/action.py:161]: Mhm, one could argue that having it after it is better for performance as the `utter_` case should be way more common.

File: rasa/core/actions/action.py (hunk 11 above)
Comment (wochinge) [reply rasa/core/actions/action.py:144]: We could shorten it to just `actions`, what do you think? I'd probably keep the old one and deprecate it's usage.
We might want to consider doing this change once everything is merged to master to not increasing the diff unnecessarily.

Comment (Ghostvv) [reply rasa/core/actions/action.py:144]: `actions` would be confusing as well, since `actions` represent actual actions `Action` class or `ActionExecuted` event

Comment (Ghostvv) [reply rasa/core/actions/action.py:144]: at this point I wouldn't worry about the diff, it is skyrocket anyhow 🙈

File: rasa/core/actions/action.py (hunk 8 above)
Comment (Ghostvv) [reply rasa/core/actions/action.py]: but what if your action_text is "utter_goodbye should be called when user says goodbye"?

**Comment by Ghostvv** [comment]
@wochinge is it ready for final review?

**Comment by wochinge** [comment]
@Ghostvv My calendar is evil today. Won't get to it until tomorrow 😬

File: rasa/core/actions/forms.py [hunk 12]
```diff
@@ -550,13 +551,16 @@ async def _ask_for_slot(
...
-        action_to_ask_for_next_slot = action.action_from_name(
-            self._name_of_utterance(domain, slot_name), domain, self.action_endpoint
-        )
-        events_to_ask_for_next_slot = await action_to_ask_for_next_slot.run(
-            output_channel, nlg, tracker, domain
-        )
-        return events_to_ask_for_next_slot
+        action_to_ask_for_next_slot = self._name_of_utterance(domain, slot_name)
+
+        if action_to_ask_for_next_slot:
```
Comment (wochinge) [inline rasa/core/actions/forms.py:556]: this change was necessary to avoid running an non-existing action (it would fail silently before)

File: rasa/core/actions/action.py [hunk 13]
```diff
@@ -147,41 +125,53 @@ def is_retrieval_action(action_name: Text, retrieval_intents: List[Text]) -> boo
...
-        name, domain.retrieval_intents
+    if action_name_or_text.startswith(UTTER_PREFIX) and is_retrieval_action(
+        action_name_or_text, domain.retrieval_intents
     ):
-        return ActionRetrieveResponse(name)
+        return ActionRetrieveResponse(action_name_or_text)
 
-    if name.startswith(UTTER_PREFIX):
-        return ActionUtterTemplate(name)
+    if action_name_or_text in domain.action_texts:
```
Comment (wochinge) [inline rasa/core/actions/action.py:160]: also added a test for this to ensure nobody thinks about switching these :-D

**Comment by wochinge** [comment]
@Ghostvv Ready for review

File: rasa/core/actions/action.py (hunk 13 above)
Comment (Ghostvv) [reply rasa/core/actions/action.py:162]: but I'd move this `if` to line 156, do you think this check will hinder the performance significantly? The win is that it'll be less error prone

**Review by Ghostvv — COMMENTED** [review]
left a comment about the order of checks. Otherwise, looks good from me. But since it introduces a lot of changes to `shared`, I think it is better if someone from backend approves it

File: rasa/core/actions/action.py (hunk 13 above)
Comment (wochinge) [reply rasa/core/actions/action.py:162]: tbh, I don't think the performance here matters too much. That should literally be a few milliseconds while e.g. some ML policies perform duplicate steps which are way more cost intense. Or we also fetch the tracker from the database more often than we need too which is in the region of ~100ms

Comment (wochinge) [reply rasa/core/actions/action.py:162]: (recently started exploring prediction performance in https://github.com/RasaHQ/rasa/pull/7205)

**Comment by wochinge** [comment]
@joejuzl Do you want to have another look or can I merge?

**Review by joejuzl — APPROVED** [review]

File: rasa/core/actions/action.py (hunk 13 above)
Comment (Ghostvv) [reply rasa/core/actions/action.py:162]: so why didn't you move it?

Comment (Ghostvv) [reply rasa/core/actions/action.py:162]: it should be the second check before `utter_` prefix check

Comment (Ghostvv) [reply rasa/core/actions/action.py:162]: @wochinge do we have a check that `action_text` is not exactly the same as some action name, then I believe will have a problem that `domain.action_name_or_text` contain non unique items

Comment (wochinge) [reply rasa/core/actions/action.py:162]: Yes, `Domain` will check it 👍  Is your message missing a `not have a problem` or is it on purpose? You mean in case a message is coming up in several stories? We de-duplicate them before ([here](https://github.com/RasaHQ/rasa/blob/49bac016e71c8a053fceeb772f0d7f8764888d07/rasa/shared/importers/importer.py#L449))

Comment (Ghostvv) [reply rasa/core/actions/action.py:162]: without the check there would be a problem, but if we have the check it should be good
